### PR TITLE
audit: do not warn about reachability of `brew mirror`ed URL

### PR DESCRIPTION
**Pull request by zmwangx**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

A `brew mirror`ed URL is usually not yet reachable at the time of pull request.

Example of a false positive fixed by this PR: https://github.com/Homebrew/homebrew-core/pull/10587.

---

**Update.** `dev-cmd/pull.rb` has been updated to warn about unreachable Bintray mirror URLs, and prompt maintainers to run `brew mirror` in that case.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
It's not reachable at PR time but it should be at every other time (including merge) so I think it should be left as-is, personally.

**Comment by ilovezfs** [comment]
Is there a way to detect if it's before or after merge?

**Comment by MikeMcQuaid** [comment]
Even before merge I personally think the CI should be red as it indicates that, as-is, it should not be merged. I appreciate it's maybe annoying running `brew audit --online` locally but I think that's better than not flagging this.

**Comment by ilovezfs** [comment]
Hmm. I think we usually do the `brew mirror` step after CI not before, since it would typically be a contributor's PR, not a maintainer's PR, so these audit failures waste maintainer and contributor time investigating them each time. It seems like it would be better to put something into `brew pull` to do this check just-in-time, and skip it during audit.

**Comment by MikeMcQuaid** [comment]
> It seems like it would be better to put something into brew pull to do this check just-in-time, and skip it during audit.

Or furthermore: just automatically do the upload. If we do this then 👍 on skipping this check on formula change PRs. I don't think it should be skipped globally as we should detect if e.g. Bintray lost some of these files.

**Comment by ilovezfs** [comment]
>Or furthermore: just automatically do the upload.

Mind. Blown.

**Comment by MikeMcQuaid** [comment]
tim-and-eric-mind-blown

**Comment by zmwangx** [comment]
> I think that's better than not flagging this.

Fair enough, is this the final verdict? If so, I'll close the PR.

> Or furthermore: just automatically do the upload.

Probably not a good idea. Too easy to submit bogus PRs to hijack a bunch of URLs.

**Comment by ilovezfs** [comment]
>Probably not a good idea. Too easy to submit bogus PRs to hijack a bunch of URLs.

I believe @MikeMcQuaid means `brew pull` should do it automatically not just random PRs.

**Comment by ilovezfs** [comment]
If @MikeMcQuaid does mean automatically doing it just because someone opened the PR, that could be problematic, but probably fine if we require the domain to match, no? Alternatively, it could be automated in some way after the PR ships.

**Comment by zmwangx** [comment]
I misread, sorry. Will take a look into adding mirror to `brew pull`.

**Comment by zmwangx** [comment]
Okay, in `pull.rb`, for each changed formula, we should be able to

1. Read the formula and find its list of URLs;
2. For each `https://dl.bintray.com/homebrew/mirror/` URL, check whether it was added by the patch by search the patch for `/^\+.*"#{Regexp.quote(url)}"`;
3. For each such URL verify whether it has a valid basename (what to do when there are multiple other URLs with different basenames?), and run `brew mirror`;
4. A `--verify` flag should probably be added to `brew mirror` to make sure the file is successfully mirrored; can be modeled on `verify_bintray_published` in `pull.rb`.

All of this should be guarded by a `--mirror` flag to `brew pull`, so maintainers still have to opt in to this, which will probably shave off much of its value.

Feels like a lot of work and a lot of added code complexity with very little return.

**Comment by ilovezfs** [comment]
Maybe we could just have `brew pull` issue a warning (or error) if the Bintray url is 404 and leave it to the maintainer to deal with?

**Comment by zmwangx** [comment]
> Maybe we could just have brew pull issue a warning (or error) if the Bintray url is 404 and leave it to the maintainer to deal with?

👍

**Comment by zmwangx** [comment]
Okay, pushed a commit to warn when a Bintray URL is unreachable. Example:

```
$ brew pull https://github.com/Homebrew/homebrew-core/commit/af44c4ba3c
Warning: Current branch is pull-test: do you need to pull inside master?
==> Fetching patch
Patch: https://github.com/Homebrew/homebrew-core/commit/af44c4ba3c.patch
/usr/bin/curl --remote-time --location --user-agent Homebrew/1.1.10-439-g475cade1a3-dirty (Macintosh; Intel macOS 10.12.3) curl/7.51.0 --fail https://github.com/Homebrew/homebrew-core/commit/af44c4ba3c.patch -s -o /Users/zmwang/Library/Caches/Homebrew/af44c4ba3c.patch
==> Applying patch
git am --whitespace=fix -3 /Users/zmwang/Library/Caches/Homebrew/af44c4ba3c.patch
Applying: imagemagick: bump homebrew mirror url for testing
Warning: The Bintray mirror https://dl.bintray.com/homebrew/mirror/imagemagick-7.0.5-404.tar.xz is not reachable (HTTP status code 404).
Warning: Do you need to upload it with `brew mirror imagemagick`?
Warning: imagemagick has a bottle: do you need to update it with --bottle?
Warning: Nonstandard bump subject: imagemagick: bump homebrew mirror url for testing
Warning: Subject should be: imagemagick 7.0.5-404
==> Patch changed:
git diff-tree -r --stat 61d247c893 HEAD
 Formula/imagemagick.rb | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
```

The relevant part:

```
Warning: The Bintray mirror https://dl.bintray.com/homebrew/mirror/imagemagick-7.0.5-404.tar.xz is not reachable (HTTP status code 404).
Warning: Do you need to upload it with `brew mirror imagemagick`?
```

File: Library/Homebrew/dev-cmd/pull.rb [hunk 1]
```diff
@@ -155,6 +155,13 @@ def pull
           next
         end
 
+        unless f.stable.nil?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pull.rb]: `if f.stable`

File: Library/Homebrew/dev-cmd/pull.rb [hunk 2]
```diff
@@ -155,6 +155,13 @@ def pull
...
+          stable_urls = [f.stable.url] + f.stable.mirrors
+          stable_urls.grep(%r{^https://dl.bintray.com/homebrew/mirror/}) do |u|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pull.rb]: `u` -> `url`

File: Library/Homebrew/dev-cmd/pull.rb [hunk 3]
```diff
@@ -588,4 +595,12 @@ def verify_bintray_published(formulae_names)
       end
     end
   end
+
+  def check_bintray_mirror(name, url)
+    headers = curl_output("--connect-timeout", "15", "--head", url)[0]
+    status_code = headers.scan(%r{^HTTP\/.* (\d+)}).last[0]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pull.rb]: `.last.first`

File: Library/Homebrew/dev-cmd/pull.rb [hunk 4]
```diff
@@ -588,4 +595,12 @@ def verify_bintray_published(formulae_names)
...
+    return if status_code.start_with?("3")
+    opoo "The Bintray mirror #{url} is not reachable (HTTP status code #{status_code})."
+    opoo "Do you need to upload it with `brew mirror #{name}`?"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pull.rb]: Use `EOS.undent` and share a `opoo`.

File: Library/Homebrew/dev-cmd/pull.rb (hunk 2 above)
Comment (zmwangx) [inline Library/Homebrew/dev-cmd/pull.rb]: Can't, shadowing outer local variable. Renamed to `mirror_url` though.

File: Library/Homebrew/dev-cmd/pull.rb (hunk 4 above)
Comment (zmwangx) [inline Library/Homebrew/dev-cmd/pull.rb]: That was the first thing I tried. It looked kinda hideous though, and the second line is easily ignored.

Nonstandard bump subject also sets the precedent for back-to-back opoos for the same issue:

```rb
          opoo "Nonstandard bump subject: #{orig_subject}"
          opoo "Subject should be: #{bump_subject}"
```

```
Warning: Nonstandard bump subject: imagemagick: bump homebrew mirror url for testing
Warning: Subject should be: imagemagick 7.0.5-404
```

**Comment by zmwangx** [comment]
@MikeMcQuaid Updated according to review comments, with one comment left unaddressed due to reason outlined above.

**Comment by MikeMcQuaid** [comment]
Thanks again @zmwangx!

**Comment by ilovezfs** [comment]
@zmwangx strange, I'm seeing
```
Josephs-MacBook-Pro:~ joe$ brew pull https://github.com/Homebrew/homebrew-core/pull/12969
==> Fetching patch 
Patch: https://github.com/Homebrew/homebrew-core/pull/12969.patch
==> Applying patch
Applying: openssl: remove support for universal build
Warning: The Bintray mirror https://dl.bintray.com/homebrew/mirror/openssl-1.0.2k.tar.gz is not reachable (HTTP status code 200).
```

**Comment by zmwangx** [comment]
This is weird...

```
$ curl -I https://dl.bintray.com/homebrew/mirror/openssl-1.0.2k.tar.gz
HTTP/1.1 200 OK
Server: nginx
Date: Sat, 29 Apr 2017 09:14:22 GMT
Content-Type: application/gzip
Content-Length: 5309236
Connection: keep-alive
Last-Modified: Sat, 29 Apr 2017 06:53:58 GMT
Accept-Ranges: none
ETag: 6b3977c61f2aedf0f96367dcfb5c6e578cf37e7b8d913b4ecb6643c3cb88d8c0
X-Checksum-Sha1: 5f26a624479c51847ebd2f22bb9f84b3b44dcb44
X-Checksum-Sha2: 6b3977c61f2aedf0f96367dcfb5c6e578cf37e7b8d913b4ecb6643c3cb88d8c0

$ curl -sSL -D- -O https://dl.bintray.com/homebrew/mirror/openssl-1.0.2k.tar.gz
HTTP/1.1 302
Server: nginx
Date: Sat, 29 Apr 2017 09:14:32 GMT
Content-Length: 0
Connection: keep-alive
Location: https://akamai.bintray.com/6b/6b3977c61f2aedf0f96367dcfb5c6e578cf37e7b8d913b4ecb6643c3cb88d8c0?__gda__=exp=1493457992~hmac=933068c51648b2e46630426957f87fb807dcb2c32608ccac0f9002d9ee0ec26b&response-content-disposition=attachment%3Bfilename%3D%22openssl-1.0.2k.tar.gz%22&response-content-type=application%2Fgzip&requestInfo=U2FsdGVkX19EwMpOdiUyB07KXMAu0IvhPWDlMRsFNUCw0jH1N3lGG8guHdJazQGFZRlCq67HdGu4ITbavFKcwn_ifBo8if-Lv7-v97cDVPHx4QdWRMOv-Y-v0ddGEN4h

HTTP/1.1 200 OK
Server: Apache
ETag: "f965fc0bf01bf882b31314b61391ae65:1491750840"
Last-Modified: Sun, 09 Apr 2017 15:09:29 GMT
Accept-Ranges: bytes
Content-Length: 5309236
Date: Sat, 29 Apr 2017 09:14:32 GMT
Connection: keep-alive
Content-Type: application/gzip
content-disposition: attachment;filename="openssl-1.0.2k.tar.gz"

$
```

Anyway, guess we should also accept 2xx status codes.

**Pull request merged**